NEW: Annotation Stores

John-P opened the pull request:
Add classes which act as a mutable mapping (dictionary like) structure and enables efficient management of annotations. An annotation here is defined as a geometry and some associated JSON data. Currently, supported features are:

- A common interface for multiple backends:
  - SQLite (with an rtree index for fast spatial queries)
  - Pure Python dict (stored as geoJSON on disk)
-  Serialisation and deserialisation to and from disk
- Compression (SQLite only)
- Conversion to and from other formats:
  - geoJSON
  - JSON Lines / ndjson
  - Pandas dataframe
- Spatial queries using a bounding box or shapely Polygon
- Customisable binary shape predicate (defaults to intersection)
- Query predicates on properties using:
  - A subset of python (fastest, compatible across backends with some limitations)
  - A pickled function
  - A python callable (slowest but easiest and can be any Python code)
- Custom indexes (SQLite only) for accelerated queries 

Example (`SQLiteStore`)
-------------------------

```python
from shapely.geometry import Polygon
from tiatoolbox.annotation.storage import SQLiteStore, Annotation

store = SQLiteStore("polygons.db")
# Create a test geometry (polygon, point, or line string)
triangle = Polygon([(0, 0), (1, 1), (0, 2)])

# Store an annotation geometry with a class label
key = store.append(Annotation(triangle, {"class": 1})) 
# The value returned in a unique key (UUID4 by default)

# Get the stored annotation (tuple of geometry and properties dict returned)
print(store[index])

# Store an annotation geometry with a custom key
store.append(Annotation(triangle, key="foo")) 
print(store["foo"])
# or use __setitem__ syntax:
store[key] = Annotation(triangle)

# Change the properties of an annotation in the store
store.patch(key, {"class": 2})

# Query in a bounding box
results = store.query([0, 0, 128, 128])

# Query using any polygon
results = store.query(Polygon.from_bounds([0, 0, 128, 128]))

# Query in a bounding box but just return only the indexes (rather than the polygons and properties)
# This can be much faster than returning geometries or doing a geometry query
# as the indexes are small and neither the geometries nor the properties have to 
# be decoded for the query to run.
results = store.query_index([0, 0, 128, 128])

# Query with a predicate statement
results = store.query([0, 0, 128, 128], where="props['class']==4")

# Create an index (SQLite only)
# Can give significant speedup (100x in some tests) even for simple property access e.g.
key = store.create_index("example_index", "props['class']==4")

```

# To-Dos

- [x] Swap to using strings for keys.
- [x] Default to UUID4 if no key given.
- [x] Allow custom string keys.
- [x] Change query shape predicate (via string kwarg).
- [x] Deepsource passing.
- [x] Query predicates.
- [x] Custom indexes.
- [x] Annotation compression (optional, defaults to zlib for sqlite).
- [x] Metadata storage (e.g. compression method, tiatoolbox version)
- [x] Test Coverage > 99%.
- [ ] Grouping during query.
- [x] Use dataclass instead of tuple for annotations?
- [x] Conform to / test Python `MutableMapping` interface
  - [x] ABC
    - [x] `__getitem__`
    - [x] `__setitem__`
    - [x] `__delitem__`
    - [x] `__iter__`
    - [x] `__len__`
  - [x] Mixins
    - [x] `__contains__`
    - [x] `keys`
    - [x] `items`
    - [x] `values`
    - [x] `get`
    - [x] `__eq__`
    - [x] `__ne__`
    - [x] `pop`
    - [x] `popitem`
    - [x] `clear`
    - [x] `update`
    - [x] `setdefault`

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1034 @@
+        - Passing None returns a zero length iterable.
+        - Passing an interable returns the passed iterable unchanged.
+        - Passing a non-iterable (e.g. and int) returns a single element iterator.
+        """
```
empty line

John-P replied:
What do you mean by this? There is an empty line at the end of this docstring

shaneahmed replied:
There should be an empty line at the end of docstring. I may have mistakenly marked it.

shaneahmed commented on tests/test_annotation_stores.py:
```diff
@@ -0,0 +1,209 @@
+        repeat_first (bool): Enforce that the last point is equal to the first.
+        direction (str): Ordering of the points. Defaults to "CCW". Valid options
+            are: counter-clockwise "CCW", and clockwise "CW".
+    """
```
empty line

shaneahmed replied:
need to add new line at the end of docstring.

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1034 @@
+        """Update an annotation at given index.
+
+        Extra key-word arguments are used to update properties.
+        """
```
empty line

vqdang commented:
There is a miscommunication between us. What I want to do is pushing, retrieval, and removal of unique instances. I didnt realize you derive the box on the fly. This resolves some of my initial confusion. Still, removing a list of UIDs is needed. 

This is what I am currently doing
```
# a dictionary contains annotations of 1 nucleus instance, the key is its unique identifier within the WSI
# take not that 1 UID => many annotations
inst_dict = {
  UUID: {
     'centroid': number[],
     'box' : number[],
     'contour' : number[],
  }
}
# adding a new instance and its associated annotations
inst_dict[uuid.uuid4()] = {
     'centroid': number[],
     'box' : number[],
     'contour' : number[],
}

# construct a separate rtree for retrieval
# inst_dict must be immutable or tree need to be recreated
boxes  = [v['box'] for v in inst_dict]
rtree = STRTree(boxes)

# retrieve any instances within bound
insts_uids = list(inst_dict.keys())
insts_in_bound = RTree.query(bound)  # pygeos returns a list of indices
remove_insts_uids = [insts_uids[idx] for idx in insts_in_bound]
# remove, and by dict association, all of their annotation are therefore removed
for uid in remove_insts_uids:
  inst_dict.pop(uid, None)
```
So I would expect the store to be a straight drop-in replacement for both the dict and STRTree.

And with the current Table form,
```
            "index": key,
            "boundary": boundary,
            "cx": int(geometry.centroid.x),
            "cy": int(geometry.centroid.y),
            "minX": geometry.bounds[0],
            "minY": geometry.bounds[1],
            "maxX": geometry.bounds[2],
            "maxY": geometry.bounds[3],
            "class": class_,
            "geom_type": geometry.geom_type,
            "properties": json.dumps(properties, separators=(",", ":")),
```
Is it possible to provide only the centroid?

John-P commented:
Ah I see, remove has not been implemented for bulk removals yet but you can use the `del` operator (`__delitem__`), the same as with dictionaries e.g.:

```python
del store[index]
```

to remove a single record.

John-P commented:
# Progress Update

I have made a few changes recently:

- Moved serialisation/deserialisation to formats such as CSV from subclasses to methods as in pandas e.g. `df.to_csv`.
- Added more functionality to act as a collection (like a dictionary) e.g. implementing `__contains__`, `.items()`, `.values()` etc.
- Simplified and cleaned up code generally.
- Moved some implementation which can be naively derived from other methods to the base class. These can be overridden with a more efficient specific implementation if required.
- Changed the update methods (`append`, `update`, `remove`) to operate on a single element and added separate bulk operation methods (`append_many`, `update_many`, `remove_many`). This simplifies implementation greatly and allows for automatic naive bulk operations if only implementing the single element operation or vice versa. The base class defines each in terms of the other and so defining either the single element or bulk operation automatically creates the other unless a better specific implementation overrides this. May have to add some check to avoid infinite recursion if neither are implemented!

John-P commented:
Now mostly working on getting test coverage up to 99%

vqdang commented:
@John-P I put here as a reminder, if it's possible, please expose query mode ('contain' vs 'intersect').

John-P commented:
Two things that I am still debating:
1. Do I add a special case for point annotations where we do not store a WKB binary for them and just use centroid x, y int columns. This would reduce storage space for points but complicate the implementation. It could also make lookups of many points a lot faster as the WKB would not need to be parsed (although this may be very small it could add up for large operations). An alternative, is to leave it simple for now and use WKB binary for everything but keep the x,y column in place (would waste a few bytes per annotation but not a big deal). Then later a special optimized point query mode could be added to just fetch many points quickly without having to decode the binary or properties at all.
2. Should I swap to using a dataclass for annotations. Currently, it is just a tuple of a geometry and properties (a dictionary). Dataclasses may be slower (and noticeable for bulk operations) but this could be mitigated by using slots.

Do you have thouts on this @vqdang  and @shaneahmed ?

shaneahmed commented on requirements.conda.yml:
```diff
@@ -24,15 +23,18 @@ dependencies:
   - scikit-image
   - scikit-learn>=0.23.2
+  - shapely
+  - shapely
```
repetition

shaneahmed commented on requirements.win64.conda.yml:
```diff
@@ -17,21 +16,24 @@ dependencies:
   - scikit-image
   - scikit-learn>=0.23.2
+  - shapely
+  - shapely
```
repetition

John-P replied:
I can't fix this currently as when I open the file there is only one ther

shaneahmed commented on tests/test_annotation_stores.py:
```diff
@@ -0,0 +1,207 @@
+) -> Polygon:
+    """Generate a fake cell boundary polygon.
+
+    Cell boundaries are generated an ellipsoids with randomised eccentricity,
```
I mean this function says cell boundaries are generated. What if you would like to generate boundaries for other structures?

John-P replied:
This is not a feature of this PR. It is just for testing purposes.

shaneahmed commented on tests/test_annotation_stores.py:
```diff
@@ -0,0 +1,980 @@
+    Adapted from pytest documentation. For more information on
+    parameterized tests see:
+    https://docs.pytest.org/en/6.2.x/example/parametrize.html#a-quick-port-of-testscenarios
+    """
```
new line at end of docstring.

shaneahmed commented on tests/test_annotation_stores.py:
```diff
@@ -0,0 +1,980 @@
+    """Test WKB deserialisation.
+
+    Test the defaul stattic method in the ABC.
+    """
```
new line at end of docstring.

shaneahmed commented on tiatoolbox/annotation/predicate.py:
```diff
@@ -0,0 +1,379 @@
+It is important to note that untrusted user input should not be
+accepted, as arbitrary code can be run during the parsing of an
+input string.
+"""
```
new line end of docstring.

shaneahmed commented on tiatoolbox/annotation/predicate.py:
```diff
@@ -0,0 +1,379 @@
+
+    Returns:
+        SQLTriplet: SQLTriplet representing None check.
+    """
```
new line end of docstring.

shaneahmed commented on tiatoolbox/annotation/predicate.py:
```diff
@@ -0,0 +1,379 @@
+
+    Returns:
+        SQLTriplet: SQLTriplet representing not None check.
+    """
```
new line end of docstring.

shaneahmed commented on tiatoolbox/annotation/predicate.py:
```diff
@@ -0,0 +1,379 @@
+
+    Returns:
+        SQLTriplet: SQLTriplet for a function call to sum the list.
+    """
```
new line end of docstring.

shaneahmed commented on tiatoolbox/annotation/predicate.py:
```diff
@@ -0,0 +1,379 @@
+
+    Returns:
+        SQLTriplet: SQLTriplet representing key check.
+    """
```
new line end of docstring.

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+        -------
+        geojson: str
+            A GeoJSON representation of this annotation.
+        """
```
new line end of docstring.

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+
+        Returns:
+            bool: True if the predicate holds.
+        """
```
new line end of docstring.

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+
+        Returns:
+            str: The unique key of the newly inserted annotation.
+        """
```
new line end of docstring.

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+
+        Returns:
+            The annotation or default if the key is not found.
+        """
```
new line end of docstring.

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+
+        Returns:
+            Any: The result of dump. Depends on the provided functions.
+        """
```
new line end of docstring.

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+        This is a naive implementation, it simply iterates over all
+        annotations and removes them. Faster implementations may be
+        possible in specific cases and may be implemented by subclasses.
+        """
```
new line end of docstring.

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+        where: Union[str, bytes, Callable[[Geometry, Dict[str, Any]], bool]] = None,
+        geometry_predicate="intersects",
+    ) -> List[str]:
+        query_geometry = geometry
```
docstring missing for this funciton.

John-P replied:
Docstring inheretence, see other comment. Can be copied if desired.

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+        where: Union[str, bytes, Callable[[Geometry, Dict[str, Any]], bool]] = None,
+        geometry_predicate: str = "intersects",
+    ) -> List[Annotation]:
+        query_geometry = geometry
```
docstring missing.

John-P replied:
Docstring should be inhereted, see other comment.

shaneahmed replied:
ok no problem. It's fine if it's inheriting it.

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+        geometries: Optional[Iterable[Geometry]] = None,
+        properties_iter: Optional[Iterable[Properties]] = None,
+    ) -> None:
+        # Validate inputs
```
docstring missing.

John-P replied:
This function should inherit the docstring of the parent classe. It can be copied out if desierd but it will be done automatically by sphinx.

shaneahmed commented on tests/test_annotation_stores.py:
```diff
@@ -0,0 +1,980 @@
+    idlist = []
+    argvalues = []
+    for scenario in metafunc.cls.scenarios:
+        idlist.append(scenario[0])
```
id_list ?

shaneahmed commented on tests/test_annotation_stores.py:
```diff
@@ -0,0 +1,980 @@
+    for scenario in metafunc.cls.scenarios:
+        idlist.append(scenario[0])
+        items = scenario[1].items()
+        argnames = [x[0] for x in items]
```
arg_names?

shaneahmed commented on tests/test_annotation_stores.py:
```diff
@@ -0,0 +1,980 @@
+    if metafunc.cls is None:
+        return
+    idlist = []
+    argvalues = []
```
arg_values ?

shaneahmed commented on tests/test_annotation_stores.py:
```diff
@@ -0,0 +1,980 @@
+# Class Specific Tests
+
+
+def test_sqlitestore_compile_options():
```
can we say sqlite_score?

John-P replied:
Why sqlite_score, the name of the class is SQLiteStore and PyTest convention is to use the lowercase in the test name.

shaneahmed commented on tests/test_annotation_stores.py:
```diff
@@ -0,0 +1,980 @@
+
+        monkeypatch.setattr(sys, "version_info", py37_version)
+        monkeypatch.setattr(sqlite3, "Connection", Connection)
+        store = store()
```
_ = store() ?

variable assigned is not used.

shaneahmed commented on tests/test_predicate.py:
```diff
@@ -0,0 +1,347 @@
+            assert isinstance(check(result), Number)
+
+    @staticmethod
+    def test_propery_binary_operations(eval_globals, eval_locals, check):
```
test_*property*_binary_operations

shaneahmed commented on tests/test_predicate.py:
```diff
@@ -0,0 +1,347 @@
+        assert check(result) == "Hello world!"
+
+    @staticmethod
+    def test_regex_ignorecase(eval_globals, eval_locals, check):
```
test_regex_ignore_case

John-P replied:
This is referring to the enum which is one word

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+    """An annotation: a geometry and associated properties.
+
+    Attributes:
+        geometry(Geometry):
```
geometry [space] (Geometry):

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+    Attributes:
+        geometry(Geometry):
+            The geometry of the annotation.
+        properties(dict):
```
properties [space] (dict):

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+    def deserialise_geometry(data: Union[str, bytes]) -> Geometry:
+        """Deserialise a geometry from a string or bytes.
+
+        This default implementaion will deserialise bytes as well-known
```
*implementation*

and *strings*

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+
+        Args:
+            data(bytes or str):
+                The serialised representaion of a Shapely geometry.
```
*representation*

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+            annotation(Annotation):
+                The shapely annotation to insert.
+            key(str):
+                Optional. The uniqure key used to identify the
```
*unique *

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+            annotations(iter(Annotation)):
+                An iterable of annotations.
+            keys(iter(str)):
+                An interable of unique keys associatetd with each
```
*iterable*
*associated*

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+
+        Returns:
+            list(str):
+                A list of unqiue keys for the inserted geometries.
```
*unique*

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+                An iterable of geometries to update.
+            properties_iter(iter(dict)):
+                An iterable of properties to update.
+            key(iter(str)):
```
keys or key? in docstring it is key and input is keys.

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+        Args:
+            key(str):
+                The key of the annoation to be removed.
+        """
```
new line end of docstring

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+            self.patch(key, geometry, properties)
+
+    def remove(self, key: str) -> None:
+        """Remove annotation from the store with its unique key.
```
*annotation*

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+            yield key
+
+    def values(self) -> Iterable[Annotation]:
+        """Return an iterable of all annotattion in the store.
```
*annotation*

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+        """Evaluate properties predicate against properties.
+
+        Args:
+            predicate(strt or bytes or Callable):
```
*str* instead of strt

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+        """Query the store for annotation keys.
+
+        Acts the same as `AnnotationStore.query` except returns keys
+        insteaf the annotations (geometry, properties).
```
instead of?

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+                For more information see the `shapely documentation on
+                binary predicates`__.
+            where:
+                A statment which should evaluate to a boolean value.
```
*statement*

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+        ]
+
+    def features(self) -> Generator[Dict[str, Any], None, None]:
+        """Return anotations as a list of geoJSON features.
```
annotations

shaneahmed commented on tiatoolbox/annotation/storage.py:
```diff
@@ -0,0 +1,1709 @@
+        """Return annotations as a dictionary in geoJSON format.
+
+        Args:
+            int_coords(bool): Make coordinates intergers. Defaults to
```
please check the input arguments.
also *integers*

shaneahmed reviewed (changes requested):
Thanks @John-P 
Please make the requested changes. we can merge then.

shaneahmed reviewed (approved):
Thanks @john
